### DISCUSSION TITLE
[FEATURE REQUEST] All file modules/states should take an encoding

### DISCUSSION THREAD

**Discussion by Timbus** [saltstack/salt#67657] [Feature Requests]
I have noticed that (on windows in particular) files with alternative encodings / utf8 BOM characters, salt's file module has a pretty rough time figuring out what to do.
`file.replace` in particular likes to delete my files when it can't handle unicode correctly which is .. unwanted. To put it politely.

**Describe the solution you'd like**
Googling around, python seems to have an excellent `chardet` module that is already a dependency in salt, so it'd be good to use that for all file operations for starters.
And in the event automatic detection isn't desirable, I think having every file function take an 'encoding' kwarg seems more than sensible. In fact, I kind of expect it..

**Comment by Timbus** [comment]
Looking into it a little more deeply, the issue I mentioned with file.replace looks to be caused by its use of mmap and lack of charset handling on write. Dunno why it doesn't just use a normal buffer, using mmap seems over the top.

This doesn't change my feature request. But it might be worth making a separate bugfix/change req for file.replace specifically, since it's deleting my files if they have a BOM.

**Comment by jtraub91** [comment]
This is cropping up for me on salt `3002.7`. Running a state file encoded with utf8 with `state.sls` on a windows minion using default cp1252 encoding errors with `UnicodeDecodeError`. Would be nice if this was either a config option or a keyword argument to pass into the necessary functions.

UPDATE: confirmed that the same issue is present on salt `3004`

**Comment by klyoungblood** [comment]
I'm seeing files removed or blanked on salt 3003.4 on Linux (CentOS7) when trying to use file.replace on files containing utf-8 encoded chars.

**Comment by smarsching** [comment]
@jtraub91 I think that the problem that you describe in your comment is actually caused by a bug in detecting the default encoding on Windows. While this is somewhat related to this feature request, I created bug report #62205 specifically for this issue (and suggested a simple fix that could be implemented by changing a single line of code). Let’s hope that we get some feedback on this from the Salt team soon.
